File: code/packages/python/virtual-memory/src/virtual_memory/tlb.py
"""TLB — Translation Lookaside Buffer.

The TLB is a small, fast cache that stores recent virtual-to-physical address
translations. Without a TLB, every memory access would require walking the
page table — which itself lives in memory. That means every single memory
access by a program would actually require 2-3 memory accesses just to find
the physical address. This would cut CPU performance by 2-3x.

The TLB fixes this by caching the most recent translations. Because programs
exhibit strong LOCALITY of reference (they tend to access the same pages over
and over), a small TLB of 32-256 entries can achieve hit rates above 95%.

How it works:
=============

    CPU wants to access virtual address 0x1ABC
        |
        v
    TLB: "Do I have VPN 0x1 cached?"
        |               |
        YES (hit!)      NO (miss)
        |               |
        v               v
    Return frame    Walk page table
    from cache      (2-3 memory accesses)
        |               |
        v               v
    Physical addr   Found frame -> cache in TLB
                        |
                        v
                    Physical addr

Why flush on context switch?
============================

When the OS switches from process A to process B, the TLB contains A's
translations. If B accesses virtual page 5, the TLB might return A's frame
for page 5 — letting B read A's private memory! Flushing the TLB on every
context switch prevents this security hole.

This is also why context switches are expensive: after a flush, the first
few memory accesses by the new process all miss the TLB and must walk the
page table. The TLB gradually "warms up" as the process runs.

Eviction Policy:
===============

When the TLB is full and a new entry needs to be inserted, we must evict
an existing entry. Our implementation uses LRU (Least Recently Used)
eviction: we track the order in which entries were accessed, and evict
the entry that hasn't been accessed for the longest time.
"""

from virtual_memory.page_table_entry import PageTableEntry


class TLB:
    """Translation Lookaside Buffer — caches VPN -> (frame, PTE) mappings.

    The TLB stores a fixed number of recent translations. On a lookup hit,
    it returns the cached physical frame number instantly. On a miss, the
    caller must walk the page table and then insert the result into the TLB.

    Attributes:
        hits: Number of successful lookups (translation was cached).
        misses: Number of failed lookups (required page table walk).

    Example:
        >>> tlb = TLB(capacity=4)
        >>> pte = PageTableEntry(frame_number=42, present=True)
        >>> tlb.insert(vpn=5, frame=42, pte=pte)
        >>> result = tlb.lookup(vpn=5)
        >>> result is not None  # hit!
        True
        >>> frame, pte = result
        >>> frame
        42
        >>> tlb.hits
        1
    """

    def __init__(self, capacity: int = 64) -> None:
        """Initialize a TLB with the given capacity.

        Args:
            capacity: Maximum number of entries the TLB can hold.
                Real TLBs have 32-256 entries. 64 is a reasonable default
                for simulation. More entries = higher hit rate but more
                expensive hardware.
        """
        # Maps VPN -> (frame_number, PTE copy).
        # We store a copy of the PTE so we can check permission bits
        # without walking the page table again.
        self._entries: dict[int, tuple[int, PageTableEntry]] = {}

        self._capacity: int = capacity

        # Track access order for LRU eviction. Most recently accessed
        # VPNs are at the END of the list. When we need to evict, we
        # remove from the FRONT (least recently used).
        self._access_order: list[int] = []

        # Statistics for measuring TLB effectiveness.
        self.hits: int = 0
        self.misses: int = 0

    def lookup(self, vpn: int) -> tuple[int, PageTableEntry] | None:
        """Look up a VPN in the TLB cache.

        If the VPN is found (a "hit"), this returns the cached frame number
        and PTE, and moves the entry to the most-recently-used position.

        If the VPN is not found (a "miss"), returns None. The caller must
        then walk the page table to find the mapping.

        Args:
            vpn: The virtual page number to look up.

        Returns:
            A tuple of (frame_number, PageTableEntry) on a hit, or None on a miss.
        """
        if vpn in self._entries:
            self.hits += 1
            # Move to end of access order (most recently used).
            # This prevents frequently-accessed pages from being evicted.
            if vpn in self._access_order:
                self._access_order.remove(vpn)
            self._access_order.append(vpn)
            return self._entries[vpn]

        self.misses += 1
        return None

    def insert(self, vpn: int, frame: int, pte: PageTableEntry) -> None:
        """Insert a translation into the TLB.

        Called after a TLB miss, when the page table walk has found the
        mapping. Caches the result so future accesses to the same VPN
        will hit the TLB instead of walking the page table.

        If the TLB is full, the least recently used entry is evicted
        to make room.

        Args:
            vpn: The virtual page number.
            frame: The physical frame number it maps to.
            pte: The page table entry (cached for permission checks).
        """
        # If already present, update in place.
        if vpn in self._entries:
            self._entries[vpn] = (frame, pte)
            if vpn in self._access_order:
                self._access_order.remove(vpn)
            self._access_order.append(vpn)
            return

        # If full, evict the least recently used entry.
        if len(self._entries) >= self._capacity:
            self._evict_lru()

        self._entries[vpn] = (frame, pte)
        self._access_order.append(vpn)

    def _evict_lru(self) -> None:
        """Evict the least recently used entry from the TLB.

        Removes the entry at the front of the access order list (the one
        that hasn't been accessed for the longest time).
        """
        if self._access_order:
            victim_vpn = self._access_order.pop(0)
            self._entries.pop(victim_vpn, None)

    def invalidate(self, vpn: int) -> None:
        """Remove a single entry from the TLB.

        Called when a specific mapping changes (e.g., after a page is
        unmapped or remapped). The stale cached translation must be removed
        so future lookups don't return the old (wrong) frame number.

        Args:
            vpn: The virtual page number to invalidate.
        """
        self._entries.pop(vpn, None)
        if vpn in self._access_order:
            self._access_order.remove(vpn)

    def flush(self) -> None:
        """Remove ALL entries from the TLB.

        Called on context switch. When the OS switches from process A to
        process B, all of A's translations become invalid because B has
        a completely different page table.

        This is expensive — the new process starts with a "cold" TLB and
        every memory access will miss until the TLB warms up. This is one
        reason why context switches are costly.
        """
        self._entries.clear()
        self._access_order.clear()

    def hit_rate(self) -> float:
        """Calculate the TLB hit rate.

        Hit rate = hits / (hits + misses)

        A good TLB has >95% hit rate. Programs exhibit temporal and spatial
        locality, meaning they repeatedly access the same few pages. A
        64-entry TLB capturing the "working set" of recent pages is enough
        for most programs.

        Returns:
            The hit rate as a float between 0.0 and 1.0, or 0.0 if no
            lookups have been performed.
        """
        total = self.hits + self.misses
        if total == 0:
            return 0.0
        return self.hits / total

    @property
    def size(self) -> int:
        """Return the current number of entries in the TLB."""
        return len(self._entries)

    @property
    def capacity(self) -> int:
        """Return the maximum capacity of the TLB."""
        return self._capacity

File: code/packages/go/virtual-memory/tlb.go
package virtualmemory

// =============================================================================
// TLB — Translation Lookaside Buffer
// =============================================================================
//
// The TLB is a small, fast cache that stores recent virtual-to-physical address
// translations. Without a TLB, every memory access would require walking the
// page table — 2-3 additional memory accesses just to find the physical address.
//
// Because programs exhibit strong LOCALITY (they access the same pages
// repeatedly), a small TLB of 32-256 entries achieves hit rates above 95%.
//
//     CPU wants to access virtual address 0x1ABC
//         |
//         v
//     TLB: "Do I have VPN 0x1 cached?"
//         |               |
//         YES (hit!)      NO (miss)
//         |               |
//         v               v
//     Return frame    Walk page table
//     from cache      (2-3 memory accesses)
//         |               |
//         v               v
//     Physical addr   Found -> cache in TLB

// tlbEntry stores a cached VPN -> (frame, PTE) mapping.
type tlbEntry struct {
	frame int
	pte   *PageTableEntry
}

// TLB caches recent virtual-to-physical translations for fast lookup.
type TLB struct {
	entries     map[int]*tlbEntry // VPN -> entry
	capacity    int
	accessOrder []int // LRU eviction: front = oldest, back = newest
	Hits        int   // Number of successful lookups
	Misses      int   // Number of failed lookups
}

// NewTLB creates a TLB with the given capacity. Real TLBs have 32-256 entries.
func NewTLB(capacity int) *TLB {
	result, _ := StartNew[*TLB]("virtual-memory.NewTLB", nil,
		func(op *Operation[*TLB], rf *ResultFactory[*TLB]) *OperationResult[*TLB] {
			op.AddProperty("capacity", capacity)
			return rf.Generate(true, false, &TLB{
				entries:     make(map[int]*tlbEntry),
				capacity:    capacity,
				accessOrder: make([]int, 0),
			})
		}).GetResult()
	return result
}

// Lookup checks the TLB for a cached translation.
//
// On a hit: returns the frame number, PTE, and true. Increments Hits.
// On a miss: returns 0, nil, and false. Increments Misses.
func (t *TLB) Lookup(vpn int) (int, *PageTableEntry, bool) {
	type lookupResult struct {
		frame int
		pte   *PageTableEntry
		ok    bool
	}
	res, _ := StartNew[lookupResult]("virtual-memory.TLB.Lookup", lookupResult{0, nil, false},
		func(op *Operation[lookupResult], rf *ResultFactory[lookupResult]) *OperationResult[lookupResult] {
			op.AddProperty("vpn", vpn)
			entry, ok := t.entries[vpn]
			if ok {
				t.Hits++
				// Move to end of access order (most recently used).
				t.removeFromOrder(vpn)
				t.accessOrder = append(t.accessOrder, vpn)
				return rf.Generate(true, false, lookupResult{entry.frame, entry.pte, true})
			}

			t.Misses++
			return rf.Generate(true, false, lookupResult{0, nil, false})
		}).GetResult()
	return res.frame, res.pte, res.ok
}

// Insert adds a translation to the TLB. If full, evicts the LRU entry.
func (t *TLB) Insert(vpn, frame int, pte *PageTableEntry) {
	_, _ = StartNew[struct{}]("virtual-memory.TLB.Insert", struct{}{},
		func(op *Operation[struct{}], rf *ResultFactory[struct{}]) *OperationResult[struct{}] {
			op.AddProperty("vpn", vpn)
			op.AddProperty("frame", frame)
			// Update existing entry.
			if _, ok := t.entries[vpn]; ok {
				t.entries[vpn] = &tlbEntry{frame: frame, pte: pte}
				t.removeFromOrder(vpn)
				t.accessOrder = append(t.accessOrder, vpn)
				return rf.Generate(true, false, struct{}{})
			}

			// Evict LRU if full.
			if len(t.entries) >= t.capacity {
				t.evictLRU()
			}

			t.entries[vpn] = &tlbEntry{frame: frame, pte: pte}
			t.accessOrder = append(t.accessOrder, vpn)
			return rf.Generate(true, false, struct{}{})
		}).GetResult()
}

// evictLRU removes the least recently used entry.
func (t *TLB) evictLRU() {
	if len(t.accessOrder) == 0 {
		return
	}
	victim := t.accessOrder[0]
	t.accessOrder = t.accessOrder[1:]
	delete(t.entries, victim)
}

// Invalidate removes a single entry from the TLB.
// Called when a specific mapping changes (e.g., page unmapped or remapped).
func (t *TLB) Invalidate(vpn int) {
	_, _ = StartNew[struct{}]("virtual-memory.TLB.Invalidate", struct{}{},
		func(op *Operation[struct{}], rf *ResultFactory[struct{}]) *OperationResult[struct{}] {
			op.AddProperty("vpn", vpn)
			delete(t.entries, vpn)
			t.removeFromOrder(vpn)
			return rf.Generate(true, false, struct{}{})
		}).GetResult()
}

// Flush removes ALL entries. Called on context switch because the new
// process has a different page table. This prevents security violations
// where one process could access another's memory via stale TLB entries.
func (t *TLB) Flush() {
	_, _ = StartNew[struct{}]("virtual-memory.TLB.Flush", struct{}{},
		func(op *Operation[struct{}], rf *ResultFactory[struct{}]) *OperationResult[struct{}] {
			t.entries = make(map[int]*tlbEntry)
			t.accessOrder = t.accessOrder[:0]
			return rf.Generate(true, false, struct{}{})
		}).GetResult()
}

// HitRate returns the ratio of hits to total lookups.
// Returns 0.0 if no lookups have been performed.
func (t *TLB) HitRate() float64 {
	result, _ := StartNew[float64]("virtual-memory.TLB.HitRate", 0,
		func(op *Operation[float64], rf *ResultFactory[float64]) *OperationResult[float64] {
			total := t.Hits + t.Misses
			if total == 0 {
				return rf.Generate(true, false, 0.0)
			}
			return rf.Generate(true, false, float64(t.Hits)/float64(total))
		}).GetResult()
	return result
}

// Size returns the current number of entries.
func (t *TLB) Size() int {
	result, _ := StartNew[int]("virtual-memory.TLB.Size", 0,
		func(op *Operation[int], rf *ResultFactory[int]) *OperationResult[int] {
			return rf.Generate(true, false, len(t.entries))
		}).GetResult()
	return result
}

// Capacity returns the maximum number of entries.
func (t *TLB) Capacity() int {
	result, _ := StartNew[int]("virtual-memory.TLB.Capacity", 0,
		func(op *Operation[int], rf *ResultFactory[int]) *OperationResult[int] {
			return rf.Generate(true, false, t.capacity)
		}).GetResult()
	return result
}

// removeFromOrder removes a VPN from the access order list.
func (t *TLB) removeFromOrder(vpn int) {
	for i, v := range t.accessOrder {
		if v == vpn {
			t.accessOrder = append(t.accessOrder[:i], t.accessOrder[i+1:]...)
			return
		}
	}
}
